### ISSUE TITLE
Restarting overwrites Nitsche particle forces and torques

### ISSUE THREAD

**Issue by anicusan** [lethe-cfd/lethe#315]
Hello, and thank you for developing this very high quality code.

## Problem Description

Once a simulation is restarted, the previously-computed nitsche particle forces and torques are lost; new values are overwritten instead of being appended to their files; e.g. I ran a simulation up to `t = 4 s`, saving forces acting on particle 0 in `solid_force_00.dat` and stopped - after restarting, `solid_force_00.dat` will only contain forces computed *after* `t = 4 s`.

I used the `gls_nitsche_navier_stokes_33` solver, though I guess file overwriting might be orthogonal to the specific solver employed. For completeness, I attach the `.prm` script I used:

```
#=================================================================================================
# Simulating two spheres approaching at constant velocity up to the point of contact, measuring
# the drag forces and torques acting thereon.
#
# The spheres' Reynolds number can be modified by changing the kinematic viscosity of the fluid
# (physical properties section); to fine-tune their individual Reynolds numbers change their radii
# and velocities (nitsche solid section).
#
# Author:   Andrei Leonard Nicusan <a.l.nicusan@bham.ac.uk>
# Date:     28.12.2021
# License:  LGPL
#=================================================================================================


#-------------------------------------------------------------------------------------------------
# Simulation and IO Control
#-------------------------------------------------------------------------------------------------
subsection simulation control
    set method                          = sdirk2
    set bdf startup method              = sdirk step

    set time step                       = 0.02                  # Time step
    set time end                        = 4.5                   # End time of simulation
    set adapt                           = true
    set max cfl                         = 0.5
    set adaptative time step scaling    = 1.1

    set output name                     = approaching_spheres   # Prefix for VTU outputs
    set output path                     = ./results/            # Output directory
    set output boundaries               = true                  # Output domain boundaries
    set output frequency                = 1                     # Frequency of simulation output
    set subdivision                     = 1                     # Output mesh subdivision
end


#-------------------------------------------------------------------------------------------------
# Simulation Checkpointing
#-------------------------------------------------------------------------------------------------
subsection restart
    set checkpoint                      = true                  # Enable checkpointing
    set restart                         = true                  # Start from previous checkpoint
    set filename                        = restart
    set frequency                       = 10
end


#-------------------------------------------------------------------------------------------------
# Force
#-------------------------------------------------------------------------------------------------
subsection forces
    set verbosity                       = verbose
    set calculate forces                = true
    set force name                      = force
    set calculate torques               = true
    set torque name                     = torque

    set output precision                = 15
    set calculation frequency           = 1
    set output frequency                = 1
end


#-------------------------------------------------------------------------------------------------
# FEM
#-------------------------------------------------------------------------------------------------
subsection FEM
    set velocity order                  = 1
    set pressure order                  = 1
    # set qmapping all                    = true
end


#-------------------------------------------------------------------------------------------------
# Physical Properties
#-------------------------------------------------------------------------------------------------
subsection physical properties
    subsection fluid 0
        set kinematic viscosity         = 0.001
    end
end


#-------------------------------------------------------------------------------------------------
# Timer
#-------------------------------------------------------------------------------------------------
subsection timer
    set type                            = iteration
end


#-------------------------------------------------------------------------------------------------
# Initial condition
#-------------------------------------------------------------------------------------------------
subsection initial conditions
    set type = nodal
    subsection uvwp
        set Function expression         = 0;0;0;0
    end
end


#-------------------------------------------------------------------------------------------------
# Mesh
#-------------------------------------------------------------------------------------------------
subsection mesh
    # Automatically generate mesh using deal.II's functions:
    # https://www.dealii.org/current/doxygen/deal.II/namespaceGridGenerator.html
    set type                            = dealii
    set grid type                       = subdivided_hyper_rectangle
    set grid arguments                  = 1,1,1 : -10,-5,-5 : 10,5,5 : true
    set initial refinement              = 3
end


#-------------------------------------------------------------------------------------------------
# Immersed Boundary Particles
#-------------------------------------------------------------------------------------------------
subsection nitsche
    set verbosity                       = verbose
    set beta                            = 10
    set calculate forces on solid       = true
    set solid force name                = solid_force
    set calculate torques on solid      = true
    set solid torque name               = solid_torque

    set number of solids                = 2

    subsection nitsche solid 0
        set enable particles motion     = true
        set particles sub iterations    = 5
        set stop if particles lost      = true

        subsection mesh
            set type                    = dealii
            set grid type               = hyper_ball
            set grid arguments          = -5,0,0 : 0.5 : true
            set initial refinement      = 5
            set simplex                 = false
        end

        subsection solid velocity
            set Function expression     = 1;0;0
        end
    end


    subsection nitsche solid 1
        set enable particles motion     = true
        set particles sub iterations    = 5
        set stop if particles lost      = true

        subsection mesh
            set type                    = dealii
            set grid type               = hyper_ball
            set grid arguments          = 5,0,0 : 0.5 : true
            set initial refinement      = 5
            set simplex                 = false
        end

        subsection solid velocity
            set Function expression     = -1;0;0
        end
    end
end


#-------------------------------------------------------------------------------------------------
# Boundary Conditions
#-------------------------------------------------------------------------------------------------
subsection boundary conditions
    # Boundaries ID 0 1 2 3 4 5 <=> xmin xmax ymin ymax zmin zmax
    # Open boundaries perpendicular to particle movement (x)
    # Slip conditions parallel to particle movement (y and z)
    set number                  = 4

    subsection bc 0
        set id                  = 2
        set type                = slip
    end

    subsection bc 1
        set id                  = 3
        set type                = slip
    end

    subsection bc 2
        set id                  = 4
        set type                = slip
    end

    subsection bc 3
        set id                  = 5
        set type                = slip
    end
end


#-------------------------------------------------------------------------------------------------
# Mesh Adaptation Control
#-------------------------------------------------------------------------------------------------
subsection mesh adaptation
    set fraction coarsening     = 0.2
    set fraction refinement     = 0.3

    set fraction type           = number
    set frequency               = 1
    set max number elements     = 1250000
    set max refinement level    = 8
    set min refinement level    = 3

    set type                    = kelly
    set variable                = pressure
end


#-------------------------------------------------------------------------------------------------
# Non-Linear Solver Control
#-------------------------------------------------------------------------------------------------
subsection non-linear solver
    set verbosity               = verbose
    set solver                  = newton
    set max iterations          = 10
    set tolerance               = 1e-5
end


#-------------------------------------------------------------------------------------------------
# Linear Solver Control
#-------------------------------------------------------------------------------------------------
subsection linear solver
    set verbosity                                   = verbose

    # GMRES linear solver, good for < 1,000,000 elements
    # set method                                      = gmres
    # set max iters                                   = 5000
    # set max krylov vectors                          = 200
    # set relative residual                           = 1e-4
    # set minimum residual                            = 1e-9
    # set ilu preconditioner fill                     = 1
    # set ilu preconditioner absolute tolerance       = 1e-12
    # set ilu preconditioner relative tolerance       = 1.00

    # AMG linear solver, more efficient for > 1,000,000 elements
    set method                                      = amg
    set max iters                                   = 5000
    set relative residual                           = 1e-4
    set minimum residual                            = 1e-9
    set amg preconditioner ilu fill                 = 1
    set amg preconditioner ilu absolute tolerance   = 1e-10
    set amg preconditioner ilu relative tolerance   = 1.00
    set amg aggregation threshold                   = 1e-12
    set amg n cycles                                = 1
    set amg w cycles                                = false
    set amg smoother sweeps                         = 2
    set amg smoother overlap                        = 1
end
```

**Comment by blaisb** [comment]
Dear Leonard,
Some could say this is a feature and not a bug :p. I'm kidding, it's a bug.
We used to not have the capabilities to reload .dat files associated with Tables, but this feature has since been written and tested. Consequently, it is an omission on our part that these data structures are not rebuilt.
We'll try to fix it in the following days / weeks. In the meantime, if you want to try your hands at it, you are more than welcomed to do so.

The feature to read a written table is tested here : https://github.com/lethe-cfd/lethe/blob/master/tests/core/table_read.cc

**Comment by jeannejoachim** [comment]
Dear Leonard,

A fix has been implemented in the pull request #349, it should be merged in the coming days. I will write again in this channel to notify you when you can update your local branch!

Thank you for reaching out, do not hesitate if you have any questions.

**Comment by jeannejoachim** [comment]
Dear Leonard,

The fix is now merged, you may update your `lethe` local repository.

Have a nice day!
